src/blockdata/script.rs

```diff
@@ -324,444 +814,505 @@ fn read_uint_iter(data: &mut core::slice::Iter<'_, u8>, size: usize) -> Result<u
-    pub fn new_witness_program(version: WitnessVersion, program: &[u8]) -> Script {
+    pub fn new_witness_program<T: AsRef<PushBytes>>(version: WitnessVersion, program: T) -> Self {
         Builder::new()
             .push_opcode(version.into())
             .push_slice(program)
             .into_script()
     }
 
     /// Generates OP_RETURN-type of scriptPubkey for the given data.
-    pub fn new_op_return(data: &[u8]) -> Script {
+    pub fn new_op_return(data: &PushBytes) -> Self {
         Builder::new()
             .push_opcode(opcodes::all::OP_RETURN)
             .push_slice(data)
             .into_script()
     }
 
-    /// Returns 160-bit hash of the script.
-    pub fn script_hash(&self) -> ScriptHash {
-        ScriptHash::hash(self.as_bytes())
-    }
-
-    /// Returns 256-bit hash of the script for P2WSH outputs.
-    pub fn wscript_hash(&self) -> WScriptHash {
-        WScriptHash::hash(self.as_bytes())
-    }
-
-    /// Returns the length in bytes of the script.
-    pub fn len(&self) -> usize { self.0.len() }
-
-    /// Returns whether the script is the empty script.
-    pub fn is_empty(&self) -> bool { self.0.is_empty() }
-
-    /// Returns the script data as a byte slice.
-    pub fn as_bytes(&self) -> &[u8] { &*self.0 }
-
-    /// Returns a copy of the script data.
-    pub fn to_bytes(&self) -> Vec<u8> { self.0.clone().into_vec() }
-
     /// Converts the script into a byte vector.
-    pub fn into_bytes(self) -> Vec<u8> { self.0.into_vec() }
+    pub fn into_bytes(self) -> Vec<u8> { self.0 }
 
     /// Computes the P2SH output corresponding to this redeem script.
-    pub fn to_p2sh(&self) -> Script {
-        Script::new_p2sh(&self.script_hash())
+    pub fn to_p2sh(&self) -> ScriptBuf {
+        ScriptBuf::new_p2sh(&self.script_hash())
     }
 
     /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
     /// for a P2WPKH output. The `scriptCode` is described in [BIP143].
     ///
     /// [BIP143]: <https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki>
-    pub fn p2wpkh_script_code(&self) -> Option<Script> {
-        if !self.is_v0_p2wpkh() {
-            return None
-        }
-        let script = Builder::new()
-            .push_opcode(opcodes::all::OP_DUP)
-            .push_opcode(opcodes::all::OP_HASH160)
-            .push_slice(&self[2..]) // The `self` script is 0x00, 0x14, <pubkey_hash>
-            .push_opcode(opcodes::all::OP_EQUALVERIFY)
-            .push_opcode(opcodes::all::OP_CHECKSIG)
-            .into_script();
-
-        Some(script)
+    pub fn p2wpkh_script_code(&self) -> Option<ScriptBuf> {
+        self.v0_p2wpkh().map(|wpkh| {
+            Builder::new()
+                .push_opcode(opcodes::all::OP_DUP)
+                .push_opcode(opcodes::all::OP_HASH160)
+                // The `self` script is 0x00, 0x14, <pubkey_hash>
```

Comment:
This is stale now, perhaps move it to the new method `v0_p2wpkh`. I realize my comment is too low level review for the current draft PR but mentioning it because what I wanted to say was I'm pretty sure I wrote this comment originally and now I can't remember why the first two bytes are 0 and 14 so if you do move the comment perhaps do a better job at explaining them than I did :)

Reply:
0x00 is witness v0
0x14 is OP_PUSHDATA (20 bytes) which is pushing the `<pubkey_hash>`

The original did `&self[2..]` to skip the 0x00 and 0x14 of the p2wpkh and only push_slice the <pubkey_hash> part. (since it's the same hash as P2PKH).

P2WPKH requires the script code preimage to be a P2PKH script for some reason (which is what this function is creating)